### ISSUE TITLE
[tune] More and more LOCAL_REFERENCE when PBT train.

### ISSUE THREAD

**Issue by noprohuangqi** [ray-project/ray#10890]
### What is your question?

When I use PBT for train my models, I use 'ray memory' and find more and more LOCAL_REFERENCE occur. And finally raise error : ObjectStoreFullError.
#### So my question is when PBT runs , whether is retain all object and never delete.For example I have 8 trails and 10 iters , is it save all 80 models as object?
  
And when I use 'reuse_actors=True', I find every 8 subprocess (8 trails) has a high memory , and when I use 'reuse_actors=False', I find every 8 subprocess (8 trails) has a low memory usage but the first start process 'python xxx.py' has  a high memory usage.
  
*Ray version and other system information (Python version, TensorFlow version, OS):*
  
Ray : 0.8.6

**Comment by amogkam** [comment]
Hi @noprohuangqi, I don't think the very high memory usage is expected. Would you be able to share what your `tune.run` call looks like as well as the code for your `Trainable`? Also would be great if you could share a screenshot of what `ray memory` looks like.

**Comment by noprohuangqi** [comment]
> Hi @noprohuangqi, I don't think the very high memory usage is expected. Would you be able to share what your `tune.run` call looks like as well as the code for your `Trainable`? Also would be great if you could share a screenshot of what `ray memory` looks like.

Thanks for your quickly reply.
This is my 'ray memory' shows(first page of 2 pages)
![image](https://user-images.githubusercontent.com/35948614/93618814-5cb1e200-fa0a-11ea-82b2-1afa9d4d9409.png)

  This is my tune.run:
 `    analysis = run(RayDemo, 
               name="ray_demo",
               checkpoint_at_end = True,
               reuse_actors=False,
               checkpoint_score_attr="bbox_mAP_50",
               checkpoint_freq=1,
#               keep_checkpoints_num=10,
#               resume = 'LOCAL',
               verbose=1,
               scheduler=pbt, 
               **train_spec)`

My Trainable is same as [this](https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/pbt_convnet_example.py)

**Comment by rkooo567** [comment]
@amogkam There's actually memory leak issue reported by me for pbt https://github.com/ray-project/ray/issues/9921

**Comment by amogkam** [comment]
Thanks @noprohuangqi. Another quick thing to try is upgrading to the latest [nightly wheel](https://docs.ray.io/en/latest/installation.html) and see if you are still getting the same issue. If you are, then I can investigate this further.

**Comment by noprohuangqi** [comment]
> Thanks @noprohuangqi. Another quick thing to try is upgrading to the latest [nightly wheel](https://docs.ray.io/en/latest/installation.html) and see if you are still getting the same issue. If you are, then I can investigate this further.

Ok. I will try the latest version as soon as posible. And other thing about this bug is when I use `reuse_actor=True`, the process  'Ray::xxx' will increase memory useage with epoch increase ; but when I use `reuse_actor=False`, the process 'Ray::xxx' will have a low and stable memory useage,but the initial process 'python start_pbt.py' will have a high memory useage.

**Comment by noprohuangqi** [comment]
> @amogkam There's actually memory leak issue reported by me for pbt #9921

Thanks for your attention.
But I do not think we are in the same bug .My failt is the more and more object, I do not know how to free those reference.

**Comment by noprohuangqi** [comment]
@amogkam 
I update the ray verison to 0.9.0, but  the question occurs again.
  
And `ray memory` shows:
![image](https://user-images.githubusercontent.com/35948614/93623244-fbd9d800-fa10-11ea-827a-d18ea16681de.png)

**Comment by noprohuangqi** [comment]
@amogkam 

Hello, this is my simplfy script to describe what quesiton I meet.
After run this file, I find `ray memory` shows Hundreds of  Reference。

```
class RayDemo(Trainable):

    def _setup(self, config):


        ## https://github.com/ray-project/ray/issues/8569
        import torch
        self.model = torch.load('resnet.pt')

    def _train(self):
        pass
        
    def _save(self, checkpoint_dir):

        file_name = 'ray.pth'
        file_path = osp.join(checkpoint_dir , file_name)
        ckpt = {'model': self.model,}

        torch.save(ckpt, file_path)
        return file_path

    def _restore(self, path):
        self.model = torch.load('resnet.pt')


    def reset_config(self, new_config):
        return True```

```
    pbt = PopulationBasedTraining(
        time_attr="training_iteration",
        metric="bbox_mAP_50",
        mode="max",
        perturbation_interval=1,
        hyperparam_mutations={
            "lr":[0,1],},
        quantile_fraction=0.5,
        resample_probability=0
        )
    
    
    analysis = run(RayDemo, 
               name="Ray_Demo",
               checkpoint_at_end = True,
               reuse_actors=False,
               checkpoint_score_attr="bbox_mAP_50",
               checkpoint_freq=1,
               verbose=1,
               scheduler=pbt, 
               **train_spec)```

**Comment by richardliaw** [comment]
Can you try again with `pip install -U ray==1.0.0rc0`?

**Comment by noprohuangqi** [comment]
> Can you try again with `pip install -U ray==1.0.0rc0`?

Thanks for your attention.
I update ray version to 1.0.0rc0, and this error occurs again.
  
This is my script `test.py`:
```#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function
import argparse
import copy
import os
import os.path as osp
import time
import random
from collections import OrderedDict
import sys
import os.path as osp
import copy
import shutil
import numpy as np
import torch
import torch.distributed as dist
import ray
from ray.tune import grid_search, run, sample_from
from ray.tune import Trainable
from ray.tune.schedulers import PopulationBasedTraining
from ray.tune.utils import validate_save_restore
import argparse
import glob
import logging
import math
import os
import random
import shutil
import time
from pathlib import Path
import torchvision

class Ray(Trainable):

    def _setup(self, config):
        ## https://github.com/ray-project/ray/issues/8569
        import torch
        self.model = torchvision.models.resnet50(pretrained=False)

    def _train(self):
        return {'bbox_mAP_50':random.random()}

    def _save(self, checkpoint_dir):

        file_name = 'ray.pth'
        file_path = osp.join(checkpoint_dir , file_name)

        ckpt = {'model': self.model,
                
                }
        torch.save(ckpt, file_path)
        return file_path

    def _restore(self, path):
        self.model = torchvision.models.resnet50(pretrained=False)

    def reset_config(self, new_config):
        return True

if __name__ == "__main__":

    train_spec = {

        "resources_per_trial": {
            "cpu": 2,
            "gpu": 1,

        },

        "stop": {
            "bbox_mAP_50": 1,
            "training_iteration": 100,
        },
        "config": {
            "lr": grid_search([0.01, 0.01*random.uniform(0.8,1.2), 0.01*random.uniform(0.8,1.2), 0.01*random.uniform(0.8,1.2),
            0.01*random.uniform(0.8,1.2), 0.01*random.uniform(0.8,1.2), 0.01*random.uniform(0.8,1.2), 0.01*random.uniform(0.8,1.2)]),
        },
        "num_samples": 1,
    }


    ray.init(dashboard_host='127.0.0.1')
    
    pbt = PopulationBasedTraining(
        time_attr="training_iteration",
        metric="bbox_mAP_50",
        mode="max",
        perturbation_interval=1,
        hyperparam_mutations={
            "lr":[0.1,0.2],},
        quantile_fraction=0.5,
        resample_probability=0
        )
    
    
    analysis = run(Ray, 
               name="ray_demo",
               checkpoint_at_end = True,
               reuse_actors=False,
               checkpoint_score_attr="bbox_mAP_50",
               checkpoint_freq=1,
               verbose=1,
               scheduler=pbt, 
               **train_spec)
```

  
And I use `top` to monitor memory usage, the process  'python test.py' has a increasing memory usage , and `ray memory` shows 216 rows of  reference object.

**Comment by richardliaw** [comment]
Maybe you could try #10905 !

**Comment by noprohuangqi** [comment]
> Maybe you could try #10905 !

Thanks for your quickly reply.
I will try this fix as soon as possible , and give you a feedback.

**Comment by richardliaw** [comment]
I think so! Did it work? @noprohuangqi

**Comment by noprohuangqi** [comment]
> I think so! Did it work? @noprohuangqi

Yes,It works. (at version 0.8.7 and 1.0.0rc0)
Thank you for your help!
